common/key_value_database/platform/lmdb/lmdb.cpp

```diff
@@ -0,0 +1,170 @@
+
+class LmdbTransaction : public Transaction {
+public:
+	LmdbTransaction(lmdb::txn &txn, lmdb::dbi &dbi);
+
+	expected::ExpectedBytes Read(const string &key) override;
+	error::Error Write(const string &key, const vector<uint8_t> &value) override;
+	error::Error Remove(const string &key) override;
+
+private:
+	lmdb::txn &txn_;
+	lmdb::dbi &dbi_;
+};
+
+LmdbTransaction::LmdbTransaction(lmdb::txn &txn, lmdb::dbi &dbi) :
+	txn_(txn),
+	dbi_(dbi) {
+}
+
+expected::ExpectedBytes LmdbTransaction::Read(const string &key) {
+	try {
+		std::string_view value;
+		bool exists = dbi_.get(txn_, key, value);
+		if (!exists) {
+			return MakeError(KeyError, "Key " + key + " not found in database");
+		}
+
+		return common::ByteVectorFromString(value);
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+}
+
+error::Error LmdbTransaction::Write(const string &key, const vector<uint8_t> &value) {
+	try {
+		string_view data(reinterpret_cast<const char *>(value.data()), value.size());
+		bool exists = dbi_.put(txn_, key, data);
+		if (!exists) {
+			return MakeError(AlreadyExistsError, "Key " + key + " already exists");
+		}
+
+		return error::NoError;
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+}
+
+error::Error LmdbTransaction::Remove(const string &key) {
+	try {
+		// We don't treat !exists as an error, just ignore return code.
+		dbi_.del(txn_, key);
+
+		return error::NoError;
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+}
+
+KeyValueDatabaseLmdb::KeyValueDatabaseLmdb() :
+	env_(lmdb::env::create()),
+	successfully_opened_(false) {
+}
+
+KeyValueDatabaseLmdb::~KeyValueDatabaseLmdb() {
+	Close();
+}
+
+error::Error KeyValueDatabaseLmdb::Open(const string &path) {
+	Close();
+
+	try {
+		env_.open(path.c_str(), MDB_NOSUBDIR, 0600);
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+
+	successfully_opened_ = true;
+	return error::NoError;
+}
+
+void KeyValueDatabaseLmdb::Close() {
+	if (!successfully_opened_) {
+		return;
+	}
+
+	env_.close();
+}
+
+expected::ExpectedBytes KeyValueDatabaseLmdb::Read(const string &key) {
+	vector<uint8_t> ret;
+	auto err = ReadTransaction([&key, &ret](Transaction &txn) -> error::Error {
+		auto result = txn.Read(key);
+		if (result) {
+			ret = std::move(result.value());
+			return error::NoError;
+		} else {
+			return result.error();
+		}
+	});
+	if (err) {
+		return err;
+	} else {
+		return ret;
+	}
+}
+
+error::Error KeyValueDatabaseLmdb::Write(const string &key, const vector<uint8_t> &value) {
+	return WriteTransaction(
+		[&key, &value](Transaction &txn) -> error::Error { return txn.Write(key, value); });
+}
+
+error::Error KeyValueDatabaseLmdb::Remove(const string &key) {
+	return WriteTransaction([&key](Transaction &txn) -> error::Error { return txn.Remove(key); });
+}
+
+error::Error KeyValueDatabaseLmdb::WriteTransaction(function<error::Error(Transaction &)> txnFunc) {
```

Comment:
The goal is that we always want the transactions to be scoped, so that you can't forget to either abort or commit. There are two ways to do this that we considered. One is the more typical C++-idiomatic way, which I actually wanted, which would look something like this:

```c++
void myfunc(KeyValueDatabase &db) {
    Transaction txn(db);
    db.Read(...stuff...);
    db.Write(...stuff...);
    // Automatically commits here, when it goes out of scope.
}
```

I started implementing it, but the problem with this approach is that committing can fail. And you are not allowed to return anything, nor even to throw exceptions, inside a destructor. So the default path would have to be to abort, and it seemed wrong to default to the failure path.

So I opted for the lambda style instead, which still keeps the scoping promise, at the cost of being a little bit more convoluted and less C++ idiomatic. You can see an example of a bigger transaction [here](https://github.com/mendersoftware/mender/blob/57a49e0fafe7d327ccee84a503083a1be923ba69/datastore/datastore.go#L98-L147), and notice that there are still no calls to `Abort()` or `Commit()`, it's all automatic based on the return value.

Reply:
But without even an option to create a transaction and then commit/abort it explicitly, I don't see the sense in using transactions in the first place. A transaction is started, some code runs, but the code doesn't manipulate the DB (What happens if it does, btw? A transaction inside a transaction?) and then the transaction is committed/aborted. What is the point of the transaction then?

Reply:
Hmm, there is some misunderstanding here, because I have to admit that I don't understand the question. Let's clear it up f2f in the standup tomorrow.